### ISSUE TITLE
Windows: UDPv4 multicast/setsockopt() for IP_ADD_MEMBERSHIP does not work

### ISSUE THREAD

**Issue by jtru** [jart/cosmopolitan#839]
I am trying to implement a minimal UDP(v4) multicast receiver loop as an actually portable executable, and cannot get the resulting binary that ought to prove this is doable to work on Windows (major version 10, x86_64, last patched today). The same binary works fine on GNU/Linux and also OS X. I am using the [**Hermetic Monorepo Snapshot**](https://justine.lol/cosmopolitan/cosmopolitan.tar.gz) to build my executable.

I am actually not sure that it is actually cosmopolitan what's somehow at fault, or if it is me (who doesn't know anything relevant about Windows programming in general, or winsock in particular) holding something wrong... The problem arises as soon as I call `setsockopt()` to make the listener join an IP multicast group by means of `IP_ADD_MEMBERSHIP` in the following code:

```c
/*
 * mcast_recv.c, massaged for cosmopolitan libc, based on
 * listener.c -- joins a multicast group and echoes all data it receives from
 *              the group to its stdout...
 *
 * Antony Courtney,     25/11/94
 * Modified by: Frédéric Bastien (25/03/04)
 * to compile without warning and work correctly
 * Modified by: Johannes Truschnigg (2023-06-25)
 * to compile into an actually portable executable (https://justine.lol/ape.html)
 */


#include "libc/calls/calls.h"
#include "libc/macros.internal.h"
#include "libc/runtime/runtime.h"
#include "libc/sock/struct/ip_mreq.h"
#include "libc/sysv/consts/so.h"
#include "libc/sysv/consts/sock.h"
#include "libc/sysv/consts/ip.h"
#include "libc/sock/sock.h"
#include "libc/sysv/consts/inaddr.h"
#include "libc/stdio/stdio.h"
#include "libc/str/str.h"
#include "libc/sysv/consts/af.h"
#include "libc/sysv/consts/ipproto.h"
#include "libc/sysv/consts/sol.h"


#define HELLO_PORT 12345
#define HELLO_GROUP "239.255.255.250"
#define MSGBUFSIZE 2048

int main(int argc, char *argv[])
{
     struct sockaddr_in addr;
     int fd, nbytes;
     uint32_t addrlen;
     struct ip_mreq mreq;
     char msgbuf[MSGBUFSIZE];

     int yes=1;

     memset(msgbuf, 0, MSGBUFSIZE);

     if ((fd=socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)) < 0) {
          perror("socket");
          exit(1);
     }

    if (setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(yes)) < 0) {
       perror("Reusing ADDR failed");
       exit(1);
       }

     memset(&addr, 0, sizeof(addr));
     addr.sin_family=AF_INET;
     addr.sin_addr.s_addr=htonl(INADDR_ANY);
     addr.sin_port=htons(HELLO_PORT);
     
     if (bind(fd, (struct sockaddr *) &addr, sizeof(addr)) < 0) {
          perror("bind");
          exit(1);
     }
     
     /* use setsockopt() to request that the kernel join a multicast group */
     mreq.imr_interface.s_addr=htonl(INADDR_ANY);
     mreq.imr_multiaddr.s_addr=inet_addr(HELLO_GROUP);
     if (setsockopt(fd, IPPROTO_IP, IP_ADD_MEMBERSHIP, &mreq, sizeof(mreq)) < 0) {
          perror("setsockopt[IP_ADD_MEMBERSHIP]");
          exit(1);
     }

     /* now just enter a read-print loop */
     addrlen=sizeof(addr);
     while (1) {
          if ((nbytes=recvfrom(fd, msgbuf, MSGBUFSIZE, 0,
                               (struct sockaddr *) &addr, &addrlen)) < 0) {
               perror("recvfrom");
               exit(1);
          }
          puts(msgbuf);
     }
}

```

When executing the resulting binary under GNU/Linux or OS X, the program enters the read-print-loop and dutifully prints any messages arriving for that multicast group to stdout. On Windows (if Defender does not interdict with some bs trojan warning :roll_eyes:), the error condition for `setsockopt()` triggers instead, and the program terminates with this error message:

`setsockopt[IP_ADD_MEMBERSHIP]: ENOPROTOOPT/10042/Protocol not available/10042/An unknown, invalid, or unsupported option or level was specified in a getsockopt or setsockopt call`

I am mainly posting this here because I don't expect anyone in some Windows programming community would know how to deal with an APE, and any MSFT docs and stackoverflow answers (pertaining to multicast network API on Windows) I found seem to suggest I am not doing anything very obviously wrong (if you disagree, **please** let me know! ;)).

Some online resources suggest that one is to use the `winsock2.h`-defined API in particular for multicast to work, but, using cosmo, I don't think that I am in a position to decide that on my own (if it's not already the case).

If anyone knows what is wrong here, and how it can be rectified to get proper UDP multicast reception working under Microsoft Windows, I would be very grateful!

**Comment by pkulchenko** [comment]
`IP_ADD_MEMBERSHIP` is mapped to system specific values based on what is [encoded in consts.sh](https://github.com/jart/cosmopolitan/blob/master/libc/sysv/consts.sh#L845).

It looks like the numbers may be wrong, as based on my search it's appears that the value should be 12 instead of 5. Can you replace the constant with 12 and see if it works for you?

There is a [similar ticket for rust](https://github.com/rust-lang/rust/issues/20381) that fixed the values according to [this change](https://github.com/rust-lang/rust/pull/21267/files).

@jart, do you want me to submit a PR to fix this?

**Comment by jtru** [comment]
Thanks for the very interesting hint, but fudging the constant in that particular way does not change the behavior on Windows for me.

For the record, *after* editing, the affected line reads like so:
```
$ awk '"IP_ADD_MEMBERSHIP" == $3' < ./libc/sysv/consts.sh
syscon  ip      IP_ADD_MEMBERSHIP                       35                      35                      12                      12                      12                      12                      12                      12                      # bsd consensus
```
This actually amounts to what you suggested me doing, right? ;)

At least https://github.com/search?q=repo%3Amingw-w64%2Fmingw-w64%20IP_ADD_MEMBERSHIP&type=code seems to hint at both 12 and/or 5 having (had?) at least some some legitimacy...

**Comment by pkulchenko** [comment]
> This actually amounts to what you suggested me doing, right? ;)

Correct.

> I am not sure if having 12 passed in does what I want, but at least there's SOME valid API/socket option hiding behind that particular constant. I will let you know if it indeed is IP_ADD_MEMBERSHIP proper once I can confirm it myself.

It would be great; thank you!

**Comment by jtru** [comment]
Hmm, is it possible that `recvfom()` is broken on win32? The POSIX manpage for the call documents that the two last args may be `NULL` (in that case, the caller doesn't care for information about the peers having put in the messages into the socket input queue, afaiui), while the [cosmopolitan libc docs for recvfrom](https://justine.lol/cosmopolitan/documentation.html#recvfrom) do not mention that possibility. Also, there is no example that actually uses `recvfrom()`, so that might have gone overlooked?


If I switch my example program to using `recv()` instead of `recvfrom()` (again afaiui, the semantic equivalent to using `recvfrom()` with `NULL` as the penultimate and last arguments), the program does not crash under Windows, while `recvfrom()`, even with properly declared/allocated arguments, always results in a crash like this:

```
recvfrom: EFAULT/487/Bad address/10014/The system detected an invalid pointer address in attempting to use a pointer argument in a call
```
... which reeks of Windows' way of telling me about a `SIGSEGV`,  I guess? :)

The `recvfrom()` implementation itself is a tad too cosmopolitan-specific and terse for my limited understanding of what going on to click, so I'd be grateful for anyone's eyes double-checking my assumption.

**Comment by pkulchenko** [comment]
> confirming that the new value for IP_ADD_MEMBERSHIP as suggested by @pkulchenko is the proper value :)

Ok, that's great; I'll check with @jart and submit a PR for it.

> The recvfrom() implementation itself is a tad too cosmopolitan-specific and terse for my limited understanding of what is going on to really click, so I'd be grateful for anyone's eyes double-checking my assumption.

There is a UDP server (tool/net/echo.c) that uses `recvfrom()` (and `sendto()`). Does that work for you or does it show the same issue you see with your app? It does fill the last two parameters though.

**Comment by jtru** [comment]
Ah, I had missed that since I was only looking in the `examples/` directory. I built the executable, and while it does work on GNU/Linux, running the UDP server variant on Windows (10 for x86_64) crashes in a very similar way my code did:

```
C:\Users\QEMU\Downloads>echoserver.com udp server 0 6000
udp server 0.0.0.0 6000
error:examples/echoserver.c:64:echoserver.com: check failed on DESKTOP-2485S1V pid 8860
        CHECK_NE(-1, (rc = recvfrom(sock, buf, sizeof(buf), 0, (struct sockaddr *)&addr2, &addrsize2)));
                 → -1 (-1)
                != -1 ((rc = recvfrom(sock, buf, sizeof(buf), 0, (struct sockaddr *)&addr2, &addrsize2)))
        EFAULT/487/Bad address/10014/The system detected an invalid pointer address in attempting to use a pointer argument in a call
        echoserver.com udp server 0 6000
note: won't print addr2line backtrace because no debug binary
70000003fe90 4048c5 NULL+0
70000003ff50 40456d NULL+0
70000003ffa0 403a4b NULL+0
70000003ffd0 403cb1 NULL+0
70000003ffe0 44e5ae NULL+0
```

So I guess this confirms something about `recvfrom()` on Windows being broken, right? Should I file a new issue, adapt this one's title, or do nothing of the sort?

**Comment by pkulchenko** [comment]
I'd file a new one, since this one is about `IP_ADD_MEMBERSHIP`; we'll take a look. Thank you!

**Comment by pkulchenko** [comment]
Microsoft also mentions this difference in [this KB article](https://learn.microsoft.com/en-US/troubleshoot/windows/win32/header-library-requirement-socket-ipproto-ip): "If the project is linked with Ws2_32.lib, setsockopt will fail with runtime error 10042 (WSAENOPROTOOPT). This is because in Winsock.h, IP_ADD_MEMBERSHIP is defined as 5. The corresponding Winsock runtime can't resolve option 5 at the IPPROTO_IP level, so the failure occurs with error code 10042."

[The options at this level are defined in Ws2tcpip.h as:](https://learn.microsoft.com/en-US/troubleshoot/windows/win32/header-library-requirement-socket-ipproto-ip#ipproto_ip-level-socket-options-in-ws2tcpiph)

```
/*
 * Option to use with [gs]etsockopt at the IPPROTO_IP level.
*/
#define IP_OPTIONS             1     /* set/get IP options */
#define IP_HDRINCL             2     /* header is included with data */
#define IP_TOS                 3     /* IP type of service and preced*/
#define IP_TTL                 4     /* IP time to live */
#define IP_MULTICAST_IF        9     /* set/get IP multicast i/f  */
#define IP_MULTICAST_TTL       10     /* set/get IP multicast ttl */
#define IP_MULTICAST_LOOP      11     /*set/get IP multicast loopback */
#define IP_ADD_MEMBERSHIP      12     /* add an IP group membership */
#define IP_DROP_MEMBERSHIP     13    /* drop an IP group membership */
#define IP_DONTFRAGMENT        14     /* don't fragment IP datagrams */
```

**Comment by pkulchenko** [comment]
Should be resolved with the closed PR #842.
